packages/reports/src/reportBuilder/fetch/fetch.ts

```diff
@@ -1,26 +1,43 @@
-import { FetchReportQuery } from '../../routes/fetchReport';
+import { FetchReportQuery } from '../../types';
 import { Aggregator } from '../../aggregator';
-import { fetchFunctions, isValidFetchFunction } from './fetchFunctions';
-import { Row } from '../reportBuilder';
+import { FetchResponse } from './types';
+import { fetchBuilders } from './functions';
 
-export type Fetch = {
-  [fetchFunctionName in keyof typeof fetchFunctions]?: object;
+type FetchParams = {
+  call: (aggregator: Aggregator, query: FetchReportQuery) => Promise<FetchResponse>;
 };
 
-export interface FetchResponse {
-  results: Row[];
-}
-
-export const fetch = (
-  fetches: Fetch[],
+const fetch = (
   aggregator: Aggregator,
   query: FetchReportQuery,
+  fetcher: FetchParams,
 ): Promise<FetchResponse> => {
-  const firstFetch: Fetch = fetches[0]; //Only support 1 fetch for now
-  const fetchFunctionKey: string = Object.keys(firstFetch)[0];
-  if (isValidFetchFunction(fetchFunctionKey)) {
-    const fetchFunction = fetchFunctions[fetchFunctionKey];
-    return fetchFunction(aggregator, query, firstFetch[fetchFunctionKey]);
+  return fetcher.call(aggregator, query);
+};
+
+const buildParams = (params: unknown): FetchParams => {
+  if (typeof params !== 'object' || params === null) {
+    throw new Error(`Expected object but got ${params}`);
```

Comment:
@kael89 yeah I totally agree. Especially since eventually we will need to do validation as part of a separate process to calculating the report (ie. when the user submits a report config).

I still like this model of doing recursive validation as it think it makes each single validation level easier to read, but yeah we should probably compartmentalise it all within a `validate` folder or something